examples/autoignition-mpi.py

```diff
@@ -244,67 +298,144 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
-            )
-            rst_data = {
-                "local_mesh": local_mesh,
-                "state": current_state,
-                "t": t,
-                "step": step,
-                "global_nelements": global_nelements,
-                "num_parts": nproc
-            }
-            from mirgecom.restart import write_restart_file
-            write_restart_file(actx, rst_data, rst_filename, comm)
-
-        # awful - computes potentially expensive viz quantities
-        #         regardless of whether it is time to viz
-        reaction_rates = eos.get_production_rates(state)
-        viz_fields = [("reaction_rates", reaction_rates)]
-        return sim_checkpoint(discr, visualizer, eos, cv=state,
-                              vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
-
-    if not check_step(current_step, nviz):  # If final step not an output step
-        if rank == 0:
-            logger.info("Checkpointing final state ...")
-        my_checkpoint(current_step, t=current_t,
-                      dt=(current_t - checkpoint_t),
-                      state=current_state)
-
-    if current_t - t_final < 0:
-        error_state = True
-
-    if error_state:
-        raise ValueError("Simulation did not complete successfully.")
+    def my_post_step(step, t, dt, state):
+        # Logmgr needs to know about EOS, dt, dim?
+        # imo this is a design/scope flaw
+        if logmgr:
+            set_dt(logmgr, dt)
+            set_sim_state(logmgr, dim, state, eos)
+            logmgr.tick_after()
+        return state, dt
+
+    def my_pre_step(step, t, dt, state):
+        dv = None
+        pre_step_errors = False
+
+        if logmgr:
+            logmgr.tick_before()
+
+        from mirgecom.simutil import check_step
+        do_viz = check_step(step=step, interval=nviz)
+        do_restart = check_step(step=step, interval=nrestart)
+        do_health = check_step(step=step, interval=nhealth)
+
+        if step == rst_step:  # don't do viz or restart @ restart
+            do_viz = False
+            do_restart = False
+
+        if do_health:
+            dv = eos.dependent_vars(state)
+            local_health_error = my_health_check(dv, dt)
+            health_errors = False
+            if comm is not None:
+                health_errors = comm.allreduce(local_health_error, op=MPI.LOR)
```

Comment:
Yes, nice catch.  I think it may be better to just rename `local_health_error` to `health_errors`.  That way we can avoid the `if` altogether.  Comm is never None anyway.

Reply:
Fixed up.